File: studio/services/selection_service.py
"""Template selection and working-html materialization."""

from studio.repositories import (
    profiles_repo,
    projects_repo,
    templates_repo,
    versions_repo,
    working_html_repo,
)


def select_template(project_id, template_id):
    if not projects_repo.get(project_id):
        raise ValueError("project not found")
    template = templates_repo.get(project_id, template_id)
    if not template:
        raise ValueError("template not found")
    html = template.get("htmlContent")
    if not html:
        raise ValueError("template html missing")
    working_html_repo.put(project_id, html, template_id=template_id)
    profile = profiles_repo.get(project_id)
    versions_repo.create(
        project_id,
        label=f"Selected {template.get('label', template_id)}",
        trigger="selection",
        html=html,
        profile=profile,
    )
    projects_repo.update_fields(
        project_id,
        {
            "status": "editing",
            "selectedTemplateId": template_id,
            "currentVersionId": versions_repo.head(project_id)["versionId"],
        },
    )
    return {
        "projectId": project_id,
        "templateId": template_id,
        "html": html,
    }

File: src/server/services/selectionService.ts
import * as profilesRepo from "../repos/profilesRepo";
import * as projectsRepo from "../repos/projectsRepo";
import * as templatesRepo from "../repos/templatesRepo";
import * as versionsRepo from "../repos/versionsRepo";
import * as workingHtmlRepo from "../repos/workingHtmlRepo";

export async function selectTemplate(
  projectId: string,
  templateId: string,
): Promise<Record<string, unknown>> {
  if (!(await projectsRepo.get(projectId))) throw new Error("project not found");
  const template = await templatesRepo.get(projectId, templateId);
  if (!template) throw new Error("template not found");
  const html = template.htmlContent;
  if (!html || typeof html !== "string") throw new Error("template html missing");
  await workingHtmlRepo.put(projectId, html, templateId);
  const profile = await profilesRepo.get(projectId);
  await versionsRepo.create(projectId, {
    label: `Selected ${template.label ?? templateId}`,
    trigger: "selection",
    html,
    profile,
  });
  const head = await versionsRepo.head(projectId);
  await projectsRepo.updateFields(projectId, {
    status: "editing",
    selectedTemplateId: templateId,
    currentVersionId: head?.versionId ?? null,
  });
  return { projectId, templateId, html };
}
